File: src/347-Top-K-Frequent-Elements.java
// given: Array of numbers, a k value
// output: An array of top k frequent elements
// constraints:  1 <= nums.length <= 10^5;

// test cases: nums = null; nums.length == 0; k > unique.values; nums.length ==1

// time and space complexity: O(n)

class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        HashMap <Integer, Integer> freqMap = new HashMap <>();

        // populate the frequency map
        for(int num: nums){
            freqMap.put(num, freqMap.getOrDefault(num,0) + 1);
        }

        // create a list of frequencies
        List<Integer>[] buckets = new ArrayList[nums.length + 1];

        // populate the buckets with values that contain the same frequencies
        for(int value: freqMap.keySet()){
            if(buckets[freqMap.get(value)] == null) buckets[freqMap.get(value)] = new ArrayList<>();
            buckets[freqMap.get(value)].add(value);
        }

        // create result array
        int[] res = new int[Math.min(k, freqMap.size())];
        int resIndex = 0;

        // populate the res array, starting with elements with the highest frequency
        for(int i = buckets.length - 1; i >= 0; i--){
            if(buckets[i] == null) continue;
            for(int element : buckets[i]){
                if(resIndex == res.length) return res;
                res[resIndex++] = element;
            }
        }

        return res;
    }


}

File: src/JavaScript-Solutions/347-Top-K-Frequent-Elements.js
/**
 * Vague questions:
 * Given an integer array nums and an integer k, return the k most frequent elements. 
 * 
 * Questions:
 * 1. Return is an array?
 * 2. Does order matter?
 * 3. What is the shortest input array that will be given?
 * 4. Is there more than one possible answer?
 * 
 * Test Cases:
 * 1. [1,1,1,1,1] k = 1
 * 2, [1,1,1,2,2,3] k = 2
 * 3. [1] k= 2
 */

/** 
347. Top K Frequent Elements

Given an integer array nums and an integer k, return the k most frequent elements. 
You may return the answer in any order.

Example 1:
  Input: nums = [1,1,1,2,2,3], k = 2
  Output: [1,2]
Example 2:
  Input: nums = [1], k = 1
  Output: [1]
 
Constraints:
  1 <= nums.length <= 105
  -104 <= nums[i] <= 104
  k is in the range [1, the number of unique elements in the array].
  It is guaranteed that the answer is unique.
*/

/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var topKFrequent = function(nums, k) {
  const result = [];
  const freqMap = new Map();
  const buckets = Array.from(Array(nums.length + 1), addEmptyArray);
  
  for (const num of nums){
      if (!freqMap.has(num)) freqMap.set(num, 0);
      freqMap.set(num, freqMap.get(num) + 1);
  }
  
  freqMap.forEach(addToBucket)
  
  for (let cur = buckets.length -1; cur > 0; cur--){
      let curBucket = buckets[cur];
      let bucketSize = curBucket.length;
      if (bucketSize === 0) continue;
      
      for (let i = 0; i < bucketSize; i++){
          if (k === 0) return result;
          result.push(curBucket[i]);
          k -= 1;
      }
  }
  
  return result;
  
  //*****************
  
  function addToBucket(freq, num){
      buckets[freq].push(num);
  }
  
  function addEmptyArray(){
      return [];
  }
  
};